### PR TITLE
PartialPageUpdater could not update if different object references were used for its own page and the component to update

### PR DESCRIPTION
**Pull request by aminabromand**
fixed equality check of PartialPageUpdater to use equals(...) method instead of equals sign

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java [hunk 1]
```diff
@@ -567,7 +567,7 @@ public final void add(final Component component, final String markupId)
 				LOG.warn("Component '{}' not cannot be updated because it was already removed from page", component);
 				return;
 			}
-			else if (pageOfComponent != page) 
+			else if (!pageOfComponent.equals(page))
```
Comment (svenmeier) [inline wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java:570]: Do you have a usecase, why this change should be necessary?

The check is precisely there to check on the page instance.

Comment (aminabromand) [reply wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java:570]: Sorry, I have missed the most important part in the pull request.
So now I have added another commit containing the `equals()` and `hashCode()` methods for the Page.class.

Well my usecase is the following:
I have a `AbstractAjaxTimerBehavior` on my page, which runs every 500 milli secs.
This `AbstractAjaxTimerBehavior` pulls Buttons out of a registry.
The registry contains the Buttons together with an information, whether the button should be enabled or disabled.
The state of the button enabled/disabled can be updated via a websocket, that is connected to the server.
So when the `AbstractAjaxTimerBehavior` pulls the Buttons our of the registry, the buttons are assigned to the same page, but it is not the very same object in the heap. 
Thus the inequality check with `pageOfComponent != page` goes true, even though the page is the same.
This is why I would like to have the inequality check as `!pageOfComponent.equals(page)` with the `equals()` method checking on the immutal fields of the page.

Do you agree, that this is a sensible approach, or would you suggest otherwise?

File: wicket-core/src/main/java/org/apache/wicket/Page.java [hunk 2]
```diff
@@ -1012,4 +1013,21 @@ public final boolean wasRendered(Component component)
...
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		Page that = (Page) o;
+		return autoIndex == that.autoIndex && numericId == that.numericId;
```
[the remark this answers is not in the document]
Comment (aminabromand) [reply wicket-core/src/main/java/org/apache/wicket/Page.java]: you are right, that `autoIndex` is not good for the method as it can change.
But so does `renderCount`.
So I change the method to use `numericId` only.

File: wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java (hunk 1 above)
Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java:570]: Changing #equals and #hashCode in such a central class can have severe negative consequence, that I would not like to risk.
Your idea with updating via component path sounds good to me.

Comment (aminabromand) [reply wicket-core/src/main/java/org/apache/wicket/page/PartialPageUpdate.java:570]: ok, understood.
Anyhow, thanks for the support and quick responses! 👍
